src/qibolab/_core/instruments/qcs.py

```diff
@@ -0,0 +1,343 @@
+        options: ExecutionParameters,
+        sweepers: list[ParallelSweepers],
+    ) -> dict[int, Result]:
+
+        if options.relaxation_time is not None:
+            self.backend._init_time = int(options.relaxation_time)
+
+        ret: dict[PulseId, np.ndarray] = {}
+        for sequence in sequences:
+            results = self.backend.apply(
+                self.create_program(
+                    sequence.align_to_delays(), configs, sweepers, options.nshots
+                )
+            ).results
+            acquisitions = sequence.acquisitions
+            acquisition_map: dict[qcs.Channels, list[InputOps]] = {}
+
+            for channel_id, input_op in acquisitions:
+                channel = self.virtual_channel_map[channel_id]
+                if channel not in acquisition_map:
+                    acquisition_map[channel] = []
+                acquisition_map[channel].append(input_op)
+
+            averaging = options.averaging_mode is not AveragingMode.SINGLESHOT
+            for channel, input_ops in acquisition_map.items():
+                if options.acquisition_type is AcquisitionType.RAW:
+                    raw = results.get_trace(channel, averaging)
+                elif options.acquisition_type is AcquisitionType.INTEGRATION:
+                    raw = results.get_iq(channel, averaging)
+                elif options.acquisition_type is AcquisitionType.DISCRIMINATION:
+                    classifier = self.classifier_map[channel]
+                    raw = results.get_classified(channel, averaging, classifier)
+                else:
+                    raise ValueError("Acquisition type unrecognized")
+
+                for result, input_op in zip(raw.values(), input_ops):
+                    # For single shot, qibolab expects result format (nshots, ...)
+                    # QCS returns (..., nshots), so we need to shuffle the arrays
+                    if (
+                        options.averaging_mode is AveragingMode.SINGLESHOT
+                        and len(sweepers) > 0
+                    ):
+                        tmp = np.zeros(options.results_shape(sweepers))
+                        if options.acquisition_type is AcquisitionType.INTEGRATION:
+                            for k in range(options.nshots):
+                                tmp[k, ..., 0] = np.real(result[..., k])
+                                tmp[k, ..., 1] = np.imag(result[..., k])
+                        else:
+                            for k in range(options.nshots):
+                                tmp[k, ...] = result[..., k]
+                        result = tmp
+
+                    elif options.acquisition_type is AcquisitionType.INTEGRATION:
+                        tmp = np.zeros(result.shape + (2,))
+                        tmp[..., 0] = np.real(result)
+                        tmp[..., 1] = np.imag(result)
+                        result = tmp
+                    ret[input_op.id] = result
```

Comment:
And here we have our last standalone function, with a docstring stub ready-made :P

Reply:
Implemented in https://github.com/qiboteam/qibolab/pull/944/commits/cf8c67e5bf62f72db50e5afebe4e53b52c14cc2f